src/engine/Union.cpp

```diff
@@ -292,3 +334,185 @@ std::unique_ptr<Operation> Union::cloneImpl() const {
+  for (auto& col : StaticRange{targetOrder_}) {
+    ColumnIndex index1 = _columnOrigins.at(col).at(0);
+    ColumnIndex index2 = _columnOrigins.at(col).at(1);
+    if (index1 == NO_COLUMN) {
+      return true;
+    }
+    if (index2 == NO_COLUMN) {
+      return false;
+    }
+    if (row1[index1] != row2[index2]) {
+      return row1[index1] < row2[index2];
+    }
+  }
+  return false;
+}
+
+// _____________________________________________________________________________
+Result::Generator Union::processRemaining(std::vector<ColumnIndex> permutation,
+                                          auto& it, auto end,
+                                          bool requestLaziness, size_t index,
+                                          IdTable& resultTable,
+                                          LocalVocab& localVocab) const {
+  // append the remaining elements
+  while (it != end) {
+    if (requestLaziness) {
+      if (index != 0) {
+        resultTable.insertAtEnd(it->idTable_, index, std::nullopt, permutation,
+                                Id::makeUndefined());
+        localVocab.mergeWith(std::span{&it->localVocab_, 1});
+        co_yield Result::IdTableVocabPair{std::move(resultTable),
+                                          std::move(localVocab)};
+      } else {
+        if (!resultTable.empty()) {
+          co_yield Result::IdTableVocabPair{std::move(resultTable),
+                                            std::move(localVocab)};
+        }
+        auto&& pair = moveOrCopy(*it);
+        pair.idTable_ = transformToCorrectColumnFormat(std::move(pair.idTable_),
+                                                       permutation);
+        co_yield pair;
+      }
+    } else {
+      resultTable.insertAtEnd(it->idTable_, index, std::nullopt, permutation,
+                              Id::makeUndefined());
+    }
+    index = 0;
+    ++it;
+  }
+}
+
+// _____________________________________________________________________________
+template <int COMPARATOR_WIDTH>
+Result::Generator Union::computeResultKeepOrderImpl(
+    bool requestLaziness, auto range1, auto range2,
+    std::pair<std::shared_ptr<const Result>, std::shared_ptr<const Result>>)
+    const {
+  constexpr size_t extent =
+      COMPARATOR_WIDTH == 0 ? std::dynamic_extent : COMPARATOR_WIDTH;
+  IdTable resultTable{getResultWidth(), allocator()};
+  if (requestLaziness) {
+    resultTable.reserve(chunkSize);
+  }
+  LocalVocab localVocab;
+  auto it1 = range1.begin();
+  auto it2 = range2.begin();
+  size_t index1 = 0;
+  size_t index2 = 0;
+  auto pushRow = [this, &resultTable](bool left, const auto& row) {
+    resultTable.emplace_back();
+    for (size_t column = 0; column < resultTable.numColumns(); column++) {
+      ColumnIndex origin = _columnOrigins.at(column).at(!left);
+      resultTable.at(resultTable.size() - 1, column) =
+          origin == NO_COLUMN ? Id::makeUndefined() : row[origin];
+    }
```

Comment:
Someone should Evalute how costly this " I am doing rowwise pushes on a columnwise data structure" is
(Is this where you spend most of the time?)
We should extend the `AddCombinedRowAdder`  to not only handle joins, but also UNIONs.